dali/imgcodec/decoders/libjpeg_turbo.cc

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/imgcodec/decoders/libjpeg_turbo.h"
+#include "dali/imgcodec/decoders/jpeg/jpeg_mem.h"
+#include "dali/imgcodec/parsers/jpeg.h"
+#include "dali/imgcodec/util/convert.h"
+
+namespace dali {
+namespace imgcodec {
+
+DecodeResult LibJpegTurboDecoderInstance::Decode(SampleView<CPUBackend> out,
+                                                 ImageSource *in,
+                                                 DecodeParams opts,
+                                                 const ROI &roi) {
+  jpeg::UncompressFlags flags;
+
+  auto &type = opts.format;
+  auto info = JpegParser{}.GetInfo(in);
+
+  flags.components = info.shape[2];
+  if (type == DALI_ANY_DATA)
+    type = info.shape[2] == 3 ? DALI_RGB : DALI_GRAY;
+
+  DALI_ENFORCE(type == DALI_RGB || type == DALI_BGR || type == DALI_GRAY,
+               "Color space not supported by libjpeg-turbo");
+  flags.color_space = type;
+
+  if (any_cast<bool>(GetParam("fast_idct"))) {
+    flags.dct_method = JDCT_FASTEST;
+  }
+
+  if (roi) {
+    flags.crop = true;
+    flags.crop_y = roi.begin[0];
+    flags.crop_x = roi.begin[1];
+    flags.crop_height = roi.shape()[0];
+    flags.crop_width = roi.shape()[1];
+  }
+
+  const uint8_t *encoded_data;
+  size_t data_size;
+  if (in->Kind() == InputKind::HostMemory) {
+    encoded_data = in->RawData<uint8_t>();
+    data_size = in->Size();
+  } else {
+    DALI_FAIL("InputKind not supported");
```

Comment:
I think it'd be good to print the input kind received (the integer representation would be enough)